File: cameraview/src/main/java/com/base/cameraview/controls/Preview.java
package com.base.cameraview.controls;

import androidx.annotation.NonNull;

/**
 * The preview engine to be used.
 */
public enum Preview implements Control {

    /**
     * Preview engine based on {@link android.view.SurfaceView}.
     * Not recommended.
     */
    SURFACE(0),

    /**
     * Preview engine based on {@link android.view.TextureView}.
     * Stable, but does not support all features (like video snapshots,
     * or picture snapshot while taking videos).
     */
    TEXTURE(1),

    /**
     * Preview engine based on {@link android.opengl.GLSurfaceView}.
     * This is the best engine available. Supports video snapshots,
     * supports picture snapshots while taking videos, supports
     * watermarks and overlays, supports real-time filters.
     */
    GL_SURFACE(2);

    final static Preview DEFAULT = GL_SURFACE;

    private int value;

    Preview(int value) {
        this.value = value;
    }

    @NonNull
    static Preview fromValue(int value) {
        Preview[] list = Preview.values();
        for (Preview action : list) {
            if (action.value() == value) {
                return action;
            }
        }
        return DEFAULT;
    }

    int value() {
        return value;
    }
}

File: cameraview/src/main/java/com/base/cameraview/controls/Preview.kt
package com.base.cameraview.controls

/**
 * The preview engine to be used.
 */
enum class Preview(val value: Int) : Control {
    /**
     * Preview engine based on [android.view.SurfaceView].
     * Not recommended.
     */
    SURFACE(0),

    /**
     * Preview engine based on [android.view.TextureView].
     * Stable, but does not support all features (like video snapshots,
     * or picture snapshot while taking videos).
     */
    TEXTURE(1),

    /**
     * Preview engine based on [android.opengl.GLSurfaceView].
     * This is the best engine available. Supports video snapshots,
     * supports picture snapshots while taking videos, supports
     * watermarks and overlays, supports real-time filters.
     */
    GL_SURFACE(2);

    companion object {
        @JvmField
        val DEFAULT: Preview = GL_SURFACE

        @JvmStatic
        fun fromValue(value: Int): Preview {
            val list = Preview.entries.toTypedArray()
            for (action in list) {
                if (action.value == value) {
                    return action
                }
            }
            return DEFAULT
        }
    }
}
